test/stats.test.js

```diff
@@ -2,83 +2,126 @@ import sinon from 'sinon';
-import { DEFAULT_OPTIONS } from '../src/consts';
+import { camelCase } from '../src/util';
 
 describe('[src/stats.js]', () => {
   let context;
   let next;
 
   beforeEach(() => {
+    sinon.stub(init, 'initStats').returns({
+      metrics: {
+        attemptCount: 0,
+        retryCount: 0,
+        requestCount: 0,
+        requestErrorCount: 0,
+        response5xxCount: 0,
+        response4xxCount: 0,
+        response3xxCount: 0,
+        response2xxCount: 0,
+        response1xxCount: 0,
+        responseInvalidCount: 0,
+        responseTime: 0,
+        cacheHitCount: 0,
+        cacheMissCount: 0,
+        cacheStaleCount: 0,
+        cacheErrorCount: 0,
+        cacheTimeoutCount: 0,
+        cacheConnectionErrorCount: 0
+      },
+      cacheAudit: []
+    });
     context = {
-      res: {}
+      cacheStatus: [],
+      res: {
+        body: 'response body',
+        headers: 'response headers',
+        statusCode: 0
+      }
     };
-    next = sinon.spy();
+    next = sinon.stub();
   });
 
   afterEach(() => sinon.restore());
 
-  describe('attempt object initialisation', () => {
-    it('should call "initAttempt" with default options', async () => {
-      sinon.stub(init, 'initAttempt').returns({});
-
-      await stats(undefined, context, next);
-
-      sinon.assert.calledOnceWithExactly(init.initAttempt, DEFAULT_OPTIONS);
-    });
-
-    it('should call "initAttempt" with the emitter if set', async () => {
-      sinon.stub(init, 'initAttempt').returns({});
-
-      await stats(DEFAULT_OPTIONS, context, next);
-
-      sinon.assert.calledOnceWithExactly(init.initAttempt, DEFAULT_OPTIONS);
-    });
-  });
+  it('should initialise the stats object and call next', async () => {
+    await stats(context, next);
 
-  it('should call next only once', async () => {
-    await stats(DEFAULT_OPTIONS, context, next);
-    sinon.assert.calledOnceWithExactly(next);
+    sinon.assert.calledOnce(init.initStats);
+    sinon.assert.calledOnce(next);
   });
 
-  describe('stats initialisation', () => {
-    it('should initialise stats if undefined', async () => {
-      const statsResponse = { attempts: [], cacheAudit: [] };
-      sinon.stub(init, 'initStats').returns(statsResponse);
-
-      await stats(DEFAULT_OPTIONS, context, next);
+  it('should throw an error with stats if "next" throws', async () => {
+    const errorMessage = 'ESOCKETTIMEDOUT';
+    next.throws(new Error(errorMessage));
 
+    try {
+      await stats(context, next);
+    } catch (error) {
       sinon.assert.calledOnce(init.initStats);
-      expect(context.res).to.be.deep.equal({ stats: statsResponse });
-    });
+      sinon.assert.calledOnce(next);
+      expect(error.message).to.be.equal(errorMessage);
+      expect(error.body).to.be.equal(context.res.body);
+      expect(error.headers).to.be.equal(context.res.headers);
+      expect(error.statusCode).to.be.equal(context.res.statusCode);
+      expect(error.stats).to.be.deep.equal({
```

Comment:
This works, of course, but it's kind of hard to see from the test what the expected change is. Perhaps we could name the object declared on L12 and describe these expected values in terms of what's changed from the initial obj?

Reply:
(actually you do this with other tests later in the file....)

Reply:
So, the object on L12 must be declared in a hook (before or after), otherwise if declared outside the describe/it scope, whenever you manipulate it, its state will persist among tests, whereas we need it to always be in that initial state (all counters to zero) and compare to the mutation afterwards. That's why it is defined directly within the stub.

Reply:
👍